source/common/filter/tcp_proxy.cc

```diff
@@ -217,29 +230,38 @@ Network::FilterStatus TcpProxy::initializeUpstreamConnection() {
       config_->stats().downstream_cx_no_route_.inc();
     }
     request_info_.setResponseFlag(AccessLog::ResponseFlag::NoRouteFound);
-    onInitFailure();
+    onInitFailure(Http::Code::ServiceUnavailable);
     return Network::FilterStatus::StopIteration;
   }
 
   Upstream::ClusterInfoConstSharedPtr cluster = thread_local_cluster->info();
   if (!cluster->resourceManager(Upstream::ResourcePriority::Default).connections().canCreate()) {
     request_info_.setResponseFlag(AccessLog::ResponseFlag::UpstreamOverflow);
     cluster->stats().upstream_cx_overflow_.inc();
-    onInitFailure();
+    onInitFailure(Http::Code::ServiceUnavailable);
     return Network::FilterStatus::StopIteration;
   }
+
+  const uint32_t max_connect_attempts = (config_ != nullptr) ? config_->maxConnectAttempts() : 1;
+  if (connect_attempts_ >= max_connect_attempts) {
+    cluster->stats().upstream_cx_connect_attempts_exceeded_.inc();
+    onInitFailure(Http::Code::GatewayTimeout);
```

Comment:
From a failure perspective I'm not sure that I would consider this a timeout technically (it would be a timeout if we actually saw timeouts when connecting), but I don't feel that strongly about it.

Reply:
I agree, but this maintained existing behavior in WS.  I'll add a comment at least.